File: qtclient/storage.cpp
/****************************************************************************
**
** Copyright (C) 2014 Digia Plc and/or its subsidiary(-ies).
** Contact: http://www.qt-project.org/legal
**
** This file is part of the examples of the Qt Toolkit.
**
** $QT_BEGIN_LICENSE:BSD$
** You may use this file under the terms of the BSD license as follows:
**
** "Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are
** met:
**   * Redistributions of source code must retain the above copyright
**     notice, this list of conditions and the following disclaimer.
**   * Redistributions in binary form must reproduce the above copyright
**     notice, this list of conditions and the following disclaimer in
**     the documentation and/or other materials provided with the
**     distribution.
**   * Neither the name of Digia Plc and its Subsidiary(-ies) nor the names
**     of its contributors may be used to endorse or promote products derived
**     from this software without specific prior written permission.
**
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
**
** $QT_END_LICENSE$
**
****************************************************************************/

#include "storage.h"
#include <QtDebug>
#include <QSettings>

#define REQUEST_URL "http://qtc-tutorial-todo.qtcloudapp.com"
#define REFRESH_INTERVAL 10000


Storage::Storage(ItemModel *model, QObject *parent) :
    QObject(parent),
    m_model(model),
    m_networkManager(0),
    m_loading(false),
    m_localIndex(0),
    m_queueLength(0)
{
    //The QNetworkAccessManager class allows the application to send network requests and receive replies
    m_networkManager = new QNetworkAccessManager(this);
    connect(m_networkManager, SIGNAL(finished(QNetworkReply*)), this, SLOT(requestFinished(QNetworkReply*)));
    deviceId = getRandomString();
    //Configure refresh timer
//    connect(&m_refreshTimer, SIGNAL(timeout()), this, SLOT(refreshItems()));
//    m_refreshTimer.setInterval(REFRESH_INTERVAL);

    //Get session information
    QSettings settings("Digia", "Qt Cloud Todo");
    m_sessionId = settings.value("sessionId", "").toByteArray();
    m_loggedName = settings.value("name", "").toString();
    m_userId = settings.value("userId", "").toString();
    m_loggedIn = !m_sessionId.isEmpty();
}

Storage::~Storage()
{
    // Save session information
    QSettings settings("Digia", "Qt Cloud Todo");
    settings.setValue("sessionId", m_sessionId);
    settings.setValue("userId", m_userId);
    settings.setValue("name", m_loggedName);
    settings.sync();
}


QNetworkReply* Storage::getWebSocketUri()
{
    QNetworkRequest request;
    request.setUrl(QUrl(QString("%1%2").arg(REQUEST_URL).arg("/api/websocket")));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    request.setRawHeader(QByteArray("x-todo-session"), m_sessionId);

    QNetworkReply *reply = m_networkManager->get(request);
    reply->setProperty("requestType", WebSocketUri);
    return reply;
}

void Storage::messageReceived(QString message) {
    qDebug() << "Message received:" << message;
    QJsonDocument jsonResponse = QJsonDocument::fromJson(message.toUtf8());
    QJsonObject jsonObj = jsonResponse.object();
    QString event = jsonObj.value("meta").toObject().value("eventName").toString();
    QJsonObject todo = jsonObj.value("object").toObject();

    if(event == "delete") {
        m_model->deleteItem(m_model->itemRow(todo.value("id").toString()));
    }
    if(todo.value("device") != deviceId) {
        if(event == "create") {
            addItem(todo.value("text").toString(), todo.value("id").toString());

        }
        else if(event == "update") {
            m_model->finishItem(m_model->itemRow(todo.value("id").toString()));
        }
    }

    setLoading(false);
}

void Storage::registerUser(const QString &name, const QString& username, const QString& password)
{
    QJsonObject data;
    data["name"] = name;
    data["username"] = username;
    data["password"] = password;

    // Send registration information to the server
    startRequest(createRequest("/api/register"), RegisterUser, true, data);
}

void Storage::loginUser(const QString& username, const QString& password)
{
    QJsonObject data;
    data["username"] = username;
    data["password"] = password;

    // Send login information to the server
    startRequest(createRequest("/api/login"), LoginUser, true, data);
}

void Storage::logoutUser()
{
    // Send logout-info to the server
    startRequest(createRequest("/api/logout", true), LogoutUser, false);

    // Logout locally
    m_refreshTimer.stop();
    m_model->clearItems();
    setLogged(false, "", "","");
}

bool Storage::rememberUser()
{
    return !m_sessionId.isEmpty();
}

void Storage::refreshItems()
{
    // Request all items from the server and update changed items to the local model. See requestFinished
    // We send refresh query only if there is no any pending requests
    if (m_queueLength == 0)
        startRequest(createRequest("/api/todos", true), RefreshItems, false);
}

void Storage::initItems()
{
    // Request all items from the server and init local model. See requestFinished
    startRequest(createRequest("/api/todos", true), InitItems);
}

void Storage::addItem(const QString& name, const QString& globalId)
{
    // Generate new localId and add item to the model with localId
    QString localId = QString("localId_%1").arg(++m_localIndex);
    m_model->addItem(localId, name, false, false);

    if(globalId == "") {
        QJsonObject data;
        data["text"] = name;
        data["done"] = false;
        data["device"] = deviceId;
        // Send request to the server
        QNetworkReply *reply = startRequest(createRequest("/api/todos", true), AddItem, false, data);
        if (reply != 0)
            reply->setProperty("localId", localId);
    }
    else {
        m_model->updateLocalId(localId, globalId);
    }
}

void Storage::finishItem(int row)
{
    // Set item in processing mode and finish it
    QString id = m_model->itemId(row);
    m_model->setProcessing(row, true);
    m_model->finishItem(row);

    // Send requeset to the server
    QJsonObject data;
    data["done"] = true;
    data["device"] = deviceId;
    startRequest(createRequest(QString("/api/todos/%1").arg(id), true), FinishItem, false, data);
}

void Storage::deleteItem(int row)
{
    // Delete item from model and send request to the server
    QString id = m_model->itemId(row);
    m_model->deleteItem(row);
    startRequest(createRequest(QString("/api/todos/%1").arg(id), true), DeleteItem, false);
}

QNetworkRequest Storage::createRequest(const QString &path, bool sessionId)
{
    QNetworkRequest request;
    request.setUrl(QUrl(QString("%1%2").arg(REQUEST_URL).arg(path)));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");

    if (sessionId)
        request.setRawHeader(QByteArray("x-todo-session"), m_sessionId);

    return request;
}

QNetworkReply* Storage::startRequest(const QNetworkRequest& request, Storage::RequestType type, bool loading, const QJsonObject& data)
{
    // Set loading flag
    setLoading(loading);

    // Send request depends on the RequestType
    QNetworkReply *reply = 0;

    switch(type) {
    case RegisterUser:
    case LoginUser:
    case AddItem:
        reply = m_networkManager->post(request, QJsonDocument(data).toJson());
        break;
    case LogoutUser:
    case InitItems:
    case RefreshItems:
        reply = m_networkManager->get(request);
        break;
    case FinishItem:
        reply = m_networkManager->put(request, QJsonDocument(data).toJson());
        break;
    case DeleteItem:
        reply = m_networkManager->deleteResource(request);
        break;
    default: break;
    }

    // If request is ok, we set requestType to the reply and handle reply in requestFinished-slot depending on this type
    if (reply != 0) {
        reply->setProperty("requestType", type);
        m_queueLength++;
    }

    return reply;
}

bool Storage::isError(const QJsonDocument& data)
{
    // Check if data includes an error
    if (data.isObject()) {
        QJsonObject d = data.object();
        if (d.contains("error")) {
            d = d.value("error").toObject();
            emit errorOccurred(d.value("code").toInt(), d.value("message").toString());
            return true;
        }
    }

    return false;
}

void Storage::requestFinished(QNetworkReply *reply)
{
    // Get requestType from the reply
    RequestType requestType = (RequestType)reply->property("requestType").toInt();
    m_queueLength--;

    // Set loading false
    setLoading(false);

    // Check reply
    if (reply->error() != QNetworkReply::NoError) {
        emit errorOccurred((int)reply->error(), reply->errorString());
    }
    else if (requestType != LogoutUser) {
        QJsonDocument replyData = QJsonDocument::fromJson(reply->readAll());
        if (!isError(replyData)) {
            switch (requestType) {
            case WebSocketUri:
            {

                QJsonObject data = replyData.object();
                m_ws_client = new WebSocketClient(QUrl(data.value("uri").toString()), this);
                QObject::connect(m_ws_client, &WebSocketClient::onMessageReceived, this, &Storage::messageReceived);
                break;
            }
            case RegisterUser:
            {
                // Send signal that user has registered
                emit userRegistered();
                break;
            }
            case LoginUser:
            {
                // Login user
                QJsonObject data = replyData.object();
                setLogged(true, data.value("name").toString(), data.value("userId").toString(), data.value("session").toString().toUtf8());
                emit userLogged();
                break;
            }
            case InitItems:
            {
                // Init all todo-items. This will occurs just after logging in
                // Start refresh timer
                //m_refreshTimer.start();
                m_model->initData(replyData.array());
                m_localIndex = m_model->rowCount() + 1;
                this->getWebSocketUri();
                break;
            }
            case RefreshItems:
            {
                // Refresh items. When user has logged in, we request/refresh items now and then (see REFRESH_INTERVAL)
                QJsonArray array = replyData.array();
                m_model->refreshData(array);
                break;
            }
            case AddItem:
            {
                // When user wants to add item, we add it to the local model immediately (see addItem())
                // Here we receive the global id to the added item and update it
                // updateLocalId-function also set processing-flag to false for this item
                m_model->updateLocalId(reply->property("localId").toString(), replyData.object().value("id").toString());
                break;
            }
            case FinishItem:
            {
                // When we receive this request, it means that item has succesfully modified in the server side.
                // So we can set processing-flag to false for this item.
                m_model->setProcessing(replyData.object().value("id").toString(), false);
                break;
            }
            default: break;
            }
        }
    }

    reply->deleteLater();
}

void Storage::setLoading(bool para)
{
    m_loading = para;
    emit loadingChanged();
}

void Storage::setLogged(bool para, const QString& name, const QString& userId, const QByteArray& sessionId)
{
    m_loggedIn = para;
    m_loggedName = name;
    m_sessionId = sessionId;
    m_userId = userId;
    emit loggedInChanged();
}

QString Storage::getRandomString() const
{
   const QString possibleCharacters("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
   const int randomStringLength = 12; // assuming you want random strings of 12 characters

   QString randomString;
   for(int i=0; i<randomStringLength; ++i)
   {
       int index = qrand() % possibleCharacters.length();
       QChar nextChar = possibleCharacters.at(index);
       randomString.append(nextChar);
   }
   return randomString;
}

File: qmlclient/qml/Storage.js
/****************************************************************************
**
** Copyright (C) 2014 Digia Plc and/or its subsidiary(-ies).
** Contact: http://www.qt-project.org/legal
**
** This file is part of the examples of the Qt Toolkit.
**
** $QT_BEGIN_LICENSE:BSD$
** You may use this file under the terms of the BSD license as follows:
**
** "Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are
** met:
**   * Redistributions of source code must retain the above copyright
**     notice, this list of conditions and the following disclaimer.
**   * Redistributions in binary form must reproduce the above copyright
**     notice, this list of conditions and the following disclaimer in
**     the documentation and/or other materials provided with the
**     distribution.
**   * Neither the name of Digia Plc and its Subsidiary(-ies) nor the names
**     of its contributors may be used to endorse or promote products derived
**     from this software without specific prior written permission.
**
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
**
** $QT_END_LICENSE$
**
****************************************************************************/

var REQUEST_URL = "http://qtc-tutorial-todo.qtcloudapp.com"
var firstDoneIndex = 0
var sessionId = ""
var localId = 0

// Show error message and logout if necessary
function error(code, message, forceLogout)
{
    if (mainView.depth > 1 && (forceLogout || code === 403))
        logoutUser()

    errorInfo.show(message === "" ? qsTr("Unknown error") : message)
}

function checkError(data)
{
    var obj = data.error
    if (obj === undefined) {
        return true;
    }
    else {
        error(obj.code, obj.message, false)
        return false;
    }
}

function initModel(data)
{
    itemModel.clear()
    firstDoneIndex = 0;
    // Add done-items to the end of the list
    for (var i=data.length; i--;) {
        var obj = data[i];
        if (obj.done)
            itemModel.append({ "itemName": obj.text, "itemDone": obj.done, "itemId": obj.id, "itemProcessing": false })
        else
        {
            itemModel.insert(0, { "itemName": obj.text, "itemDone": obj.done, "itemId": obj.id, "itemProcessing": false })
            firstDoneIndex++;
        }
    }
}

function refreshModel(data)
{
    // 1. Update available items and delete unnecessary items
    for (var i=itemModel.count; i--;) {
        var id = itemModel.get(i).itemId
        var found = false;
        for (var j=data.length; j--;) {
            var obj = data[j];
            if (obj.id === id) {
                itemModel.setProperty(i, "itemName", obj.text);
                if (itemModel.get(i).itemDone !== obj.done)
                {
                    itemModel.setProperty(i, "itemDone", obj.done);
                    if (i !== firstDoneIndex && i !== (firstDoneIndex-1))
                        itemModel.move(i,firstDoneIndex-1,1)

                    if (obj.done)
                        firstDoneIndex--;
                    else
                        firstDoneIndex++;
                }

                data.splice(j,1);
                found = true;
                break;
            }
        }

        if (!found)
            itemModel.remove(i);
    }

    // 2. Add new items
    for (var k=0; k<data.length; k++) {
        itemModel.insert(0, { "itemName": data[k].text, "itemDone": data[k].done, "itemId": data[k].id, "itemProcessing": false })
        firstDoneIndex++;
    }

    // 3. Calculate first done-index
    firstDoneIndex = itemModel.count;
    for (var l=0; l<itemModel.count; l++) {
        if (itemModel.get(l).itemDone) {
            firstDoneIndex = l;
            break;
        }
    }
}

// HTTP-functions -->
function registerUser(name, username, password)
{
    app.loading = true

    var doc = new XMLHttpRequest();
    doc.open("POST", REQUEST_URL + "/api/register", true);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.onreadystatechange = function()
    {
        // When ready
        if (doc.readyState === 4) {

            // If OK
            if (doc.status === 200) {
                var data = JSON.parse(doc.responseText);
                if (checkError(data))
                    loginUser(username, password);  // Login after registering
            }
            else {
                error(doc.status, doc.statusText, true)
            }
        }
    }
    doc.send(JSON.stringify({ "name": name, "username": username, "password": password }));
}

function loginUser(username, password)
{
    app.loading = true

    var doc = new XMLHttpRequest();
    doc.open("POST", REQUEST_URL + "/api/login", true);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.onreadystatechange = function()
    {
        // When ready
        if (doc.readyState === 4) {

            // If OK
            if (doc.status === 200) {
                var data = JSON.parse(doc.responseText);
                if (checkError(data)) {
                    Qt.inputMethod.hide()
                    mainView.push({ item: Qt.resolvedUrl("Todo.qml"), replace: true });
                    app.loggedIn = true
                    app.loggedName = data.name
                    sessionId = data.session
                    // Open WebSocket connection
                    connectToWebSocket()
                    // Init items after login
                    initItems()
                }
            }
            else {
                error(doc.status, doc.statusText, true)
            }
        }
    }
    doc.send(JSON.stringify({ "username": username, "password": password }));
}

function connectToWebSocket() {
    var doc = new XMLHttpRequest();
    doc.open("GET", REQUEST_URL + "/api/websocket", true);
    doc.setRequestHeader('x-todo-session', sessionId);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.onreadystatechange = function()
    {
        // When ready
        if (doc.readyState === 4) {

            // If OK
            if (doc.status === 200) {
                var data = JSON.parse(doc.responseText);
                console.log("Open WebSocket connection", data.uri)
                socket.url = data.uri
            }
        }
    }
    doc.send()
}

function logoutUser()
{

    // Pop pages
    while (mainView.depth > 1)
        mainView.pop()

    app.loading = false

    // Hide keyboard
    Qt.inputMethod.hide()

    // Send logout-info to the server
    var doc = new XMLHttpRequest();
    doc.open("GET", REQUEST_URL + "/api/logout", true);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.send();
}

function initItems()
{
    var doc = new XMLHttpRequest();
    doc.open("GET", REQUEST_URL + "/api/todos", true);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.setRequestHeader('x-todo-session', sessionId);
    doc.onreadystatechange = function()
    {
        // When ready
        if (doc.readyState === 4) {

            // If OK
            if (doc.status === 200) {
                var data = JSON.parse(doc.responseText);
                if (checkError(data)) {
                    initModel(data)
                    localId = itemModel.count + 1
                    app.loading = false
                }
            }
            else {
                error(doc.status, doc.statusText, false)
            }
        }
    }
    doc.send();
}

function refreshItems()
{
    var doc = new XMLHttpRequest();
    doc.open("GET", REQUEST_URL + "/api/todos", true);
    doc.setRequestHeader('Content-Type', 'application/json');
    doc.setRequestHeader('x-todo-session', sessionId);
    doc.onreadystatechange = function()
    {
        // When ready
        if (doc.readyState === 4) {

            // If OK
            if (doc.status === 200) {
                var data = JSON.parse(doc.responseText);
                if (checkError(data))
                    refreshModel(data);
            }
            else {
                error(doc.status, doc.statusText, false)
            }
        }
    }
    doc.send();
}

function addItem(name, itemId)
{
    console.log('Create item '+name)
    var locId = itemId || "local_" + (localId++);
    var processing = itemId ? false : true
    // Add item locally
    itemModel.insert(0, { "itemName": name, "itemDone": false, "itemId": locId, "itemProcessing": processing  })
    firstDoneIndex++;
    // if itemId not provided sync to backend
    if(!itemId){
        // Send info to the server
        var doc = new XMLHttpRequest();
        doc.open("POST", REQUEST_URL + "/api/todos", true);
        doc.setRequestHeader('Content-Type', 'application/json');
        doc.setRequestHeader('x-todo-session', sessionId);
        doc.onreadystatechange = function()
        {
            // When ready
            if (doc.readyState === 4) {

                // If OK
                if (doc.status === 200) {
                    var data = JSON.parse(doc.responseText);
                    if (checkError(data)) {
                        for (var i=0; i<itemModel.count; i++) {
                            if (itemModel.get(i).itemId === locId) {
                                itemModel.setProperty(i, "itemId", data.id);
                                itemModel.setProperty(i, "itemProcessing", false)
                                break;
                            }
                        }
                    }
                    else {
                        error(doc.status, doc.statusText, false)
                    }
                }
            }
        }

        doc.send(JSON.stringify({ "text": name, "done": false, "device": device }));
    }
}

function finishItem(row, remote)
{
    console.log('Complete item')
    var processing = remote ? true : false
    if (row >= 0 && row < itemModel.count)
    {
        // Set item finished locally
        itemModel.setProperty(row, "itemDone", true);
        itemModel.setProperty(row, "itemProcessing", processing);
        var obj = itemModel.get(row)
        if (row !== firstDoneIndex && row !== (firstDoneIndex-1))
            itemModel.move(row,firstDoneIndex-1,1)
        firstDoneIndex--;
        if(remote === true) {
            // Send information to the server
            var doc = new XMLHttpRequest();
            doc.open("PUT", REQUEST_URL + "/api/todos/"+obj.itemId, true);
            doc.setRequestHeader('Content-Type', 'application/json');
            doc.setRequestHeader('x-todo-session', sessionId);
            doc.onreadystatechange = function()
            {
                // When ready
                if (doc.readyState === 4) {

                    // If OK
                    if (doc.status === 200) {
                        var data = JSON.parse(doc.responseText);
                        if (checkError(data)) {
                            for (var i=0; i<itemModel.count; i++) {
                                if (itemModel.get(i).itemId === data.id) {
                                    itemModel.setProperty(i, "itemProcessing", false)
                                    break;
                                }
                            }
                        }
                    }
                    else {
                        error(doc.status, doc.statusText, false)
                    }
                }
            }
            doc.send(JSON.stringify({ "text": obj.itemName, "done": true, "device": device }));
        }
    }
}

function deleteItem(row, remote)
{
    console.log("Delete item")
    if (row >= 0 && row < itemModel.count)
    {
        // Delete locally
        var id = itemModel.get(row).itemId;
        itemModel.remove(row);

        if(remote === true) {
            // Send info to server
            var doc = new XMLHttpRequest();
            doc.open("DELETE", REQUEST_URL + "/api/todos/"+id, true);
            doc.setRequestHeader('Content-Type', 'application/json');
            doc.setRequestHeader('x-todo-session', sessionId);
            doc.send();
        }
    }
}

function handleWebsocketMessage(message) {
    var messageJson = JSON.parse(message)
    var event = messageJson.meta.eventName
    var object = messageJson.object
    console.log("Message received", event, object.text);
    console.log("Message payload:" ,message);

    if(object.device !== device) { // ignore events created by this app
        var i
        if(event === "create") {
            addItem(object.text, object.id)
        }
        else if(event === "update") {
            for (i=itemModel.count; i--;) {
                if( object.id === itemModel.get(i).itemId) {
                    finishItem(i, false)
                    break;
                }
            }
        }
    }
    if(event === "delete") {
        for (i=itemModel.count; i--;) {
            if( object.id === itemModel.get(i).itemId) {
                deleteItem(i, false)
                break;
            }
        }
    }
}
